### PR TITLE
[jit] make clone works for interface type

### LINKED ISSUE #41882: [JIT] Module cloning fails with certain hierarchies involving module interfaces
## To Reproduce
**Python**
```
import torch
import torch.nn as nn

        
@torch.jit.interface
class OperatorIf(nn.Module):
    def forward(self, inp: torch.Tensor) -> torch.Tensor:
        pass


class Operator(nn.Module):
    def __init__(self, a):
        super().__init__()
        self.a = a
        
    def forward(self, inp: torch.Tensor) -> torch.Tensor:
        return self.a * (inp + self.a)
        

class Inner(nn.Module):
    op: OperatorIf
    def __init__(self, op):
        super().__init__()
        self.op = op
        
    def forward(self, inp):
        return self.op(inp)


class Outer(nn.Module):
    def __init__(self):
        super().__init__()
        self.inner_a = Inner(Operator(1))
        self.inner_b = Inner(Operator(3.0))


if __name__ == "__main__":
    s = torch.jit.script(Outer())
    torch.jit.save(s, "/tmp/s.pt")
```

**C++**
```
  std::ifstream in("/tmp/s.pt");
  auto m = torch::jit::load(in);
  auto cloned = m.clone();
```
## Expected behavior
Module clones successfully or a clear error message is shown explaining the problem.

## Actual behaviour
Exception is thrown:
```
terminate called after throwing an instance of 'c10::Error'
  what():  __torch__.___torch_mangle_3.Operator is not compatible with the type __torch__.___torch_mangle_2.Operator for the field 'op'
```

cc @ezyang @gchanan @zou3519 @suo @gmagogsfm

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42121 [jit] make clone works for interface type**

This PR changes the Module clone to allow clone a module with module
interface type, to allow Module::clone works on the case
where there's a module interface type being shared by two submodules.

interface type will be shared by the new cloned instance in the same
compilation unit bc it only
contains a list of functionSchema, which does not involve any
attributes compared to classType.

fixes https://github.com/pytorch/pytorch/issues/41882

Differential Revision: [D22781205](https://our.internmc.facebook.com/intern/diff/D22781205)

### CODE REVIEW

File: torch/csrc/jit/api/module.h [hunk 1]
```diff
@@ -141,8 +141,35 @@ struct TORCH_API Module : public Object {
     _ivalue()->setAttr(name, std::move(v));
   }
 
-  void register_module(const std::string& name, const Module& module) {
-    type()->addOrCheckAttribute(name, module.type());
+  void register_module(
+      const std::string& name,
+      const Module& module,
```
Comment (suo) [inline torch/csrc/jit/api/module.h:146]: It seems to me that this is a special case (where we want to add a module attribute but register its type as an interface). My suggestion is to avoid changing the public module API and do this in the implementation of cloning, to hide this from users. We already have special rules for module attributes there, so we can just change that rule to do:
1. Copy the attribute type
2. Insert the cloned submodule into the ivalue.

File: torch/csrc/jit/api/module.cpp [hunk 2]
```diff
@@ -203,18 +203,22 @@ Module Module::clone_impl(
...
   for (size_t i = 0; i < N; ++i) {
     IValue s = _ivalue()->getSlot(i);
-    if (type()->getAttribute(i)->is_module()) {
+    TypePtr attr_type = type()->getAttribute(i);
+    if (attr_type->is_module()) {
       const Module& orig = Module(s.toObject());
       Module cloned = orig.clone_impl(type_remap, inplace, memo);
       type_remap[orig.type()] = cloned.type();
-      r.register_module(type()->getAttributeName(i), cloned);
+      r.register_module(
```
Comment (suo) [inline torch/csrc/jit/api/module.cpp:211]: Concretely, we should change this from a call to `register_module` to (pseudocode):
```
r.type.addAttribute(name, interface)
r.ivalue.setattr(name, cloned_submodule)
```

**Review by suo — CHANGES_REQUESTED** [review]

**Comment by SplitInfinity** [comment]
I patched this and I still see the same error on the test case that I included in #41882.

**Comment by wanchaol** [comment]
> I patched this and I still see the same error on the test case that I included in #41882.

hmm are you sure? I manually tried the example you had in the issue (add the C++ code in the issue to `testModuleCloneWithModuleInterface`, and it works fine on my side.

File: torch/csrc/jit/api/module.h (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/jit/api/module.h:146]: sounds good, I will change it accordingly.

**Comment by SplitInfinity** [comment]
Okay I will try again

**Comment by jerryzh168** [comment]
Thanks, we also have a internal clone in quantization: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#_ZN5torch3jit12_GLOBAL__N_117ModuleCloneHelper5cloneERKNS0_6ModuleERKSt13unordered_mapIN3c1013intrusive_ptrINS7_6ivalue6ObjectENS7_6detail34intrusive_14255110, could you change that as well?

this is for unsharing the types that are configured with different qconfig.

**Comment by SplitInfinity** [comment]
Okay, I made a mistake. I forgot to patch the diff in fbcode before trying. It works.

**Comment by wanchaol** [comment]
> Thanks, we also have a internal clone in quantization: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#_ZN5torch3jit12_GLOBAL__N_117ModuleCloneHelper5cloneERKNS0_6ModuleERKSt13unordered_mapIN3c1013intrusive_ptrINS7_6ivalue6ObjectENS7_6detail34intrusive_14255110, could you change that as well?
> 
> this is for unsharing the types that are configured with different qconfig.

@jerryzh168 thanks for the pointer! I will update. One question, do you know where I should add the test for the quantization internal clone impl?

File: torch/csrc/jit/api/module.h (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/jit/api/module.h:146]: I took a second thought on this. Right now in python, we have a way for user to define an interface attr with a submodule, but we don't have a equivalent way to do this in C++. Does it mean that right now if we do it in C++, user have to call `register_attribute` for a module interface type? that seems a bit weird to me.

**Comment by jerryzh168** [comment]
> > Thanks, we also have a internal clone in quantization: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#_ZN5torch3jit12_GLOBAL__N_117ModuleCloneHelper5cloneERKNS0_6ModuleERKSt13unordered_mapIN3c1013intrusive_ptrINS7_6ivalue6ObjectENS7_6detail34intrusive_14255110, could you change that as well?
> > this is for unsharing the types that are configured with different qconfig.
> 
> @jerryzh168 thanks for the pointer! I will update. One question, do you know where I should add the test for the quantization internal clone impl?

I think you can have two tests for this:
1. The code is called on all quantization invocations, so you can test it by quantizing any model with interface type.

2. To test unsharing, I think we can do following:
create a model with interface types and two sub modules both have the interface type. and configure the two sub modules with the interface type with different qconfigs, and then quantize it.

copying test case from https://github.com/pytorch/pytorch/issues/41882
```
import torch
import torch.nn as nn

        
@torch.jit.interface
class OperatorIf(nn.Module):
    def forward(self, inp: torch.Tensor) -> torch.Tensor:
        pass


class Operator(nn.Module):
    def __init__(self, a):
        super().__init__()
        self.a = a
        
    def forward(self, inp: torch.Tensor) -> torch.Tensor:
        return self.a * (inp + self.a)
        

class Inner(nn.Module):
    op: OperatorIf
    def __init__(self, op):
        super().__init__()
        self.op = op
        
    def forward(self, inp):
        return self.op(inp)


class Outer(nn.Module):
    def __init__(self):
        super().__init__()
        self.inner_a = Inner(Operator(1))
        self.inner_b = Inner(Operator(3.0))

qconfig_dict = {'inner_a': default_qconfig, 'inner_b': default_qconfig}
model = torch.jit.script(Outer()) # maybe add a trace test as well
# make sure it runs
prepare_jit(...)
```
This should be added here: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_jit.py#L381

e.g. `test_insert_observers_interface` and `test_insert_observers_interface_unshare_type`

File: torch/csrc/jit/api/module.h (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/api/module.h]: I think we don't intend for people to construct modules with interface types directly from C++ (for example, the C++ constructors for interfaces are not part of the public API in any way). If we want to add that functionality, we would need to add a lot more than just the bindings on `Module`, so my preference is to leave it out for now.

File: test/quantization/test_quantize_jit.py [hunk 3]
```diff
@@ -395,6 +395,79 @@ def forward(self, x):
...
+
+            def forward(self, inp):
+                return self.inner_a(inp) + self.inner_b(inp)
+
+        qconfig_dict = {'inner_a': default_qconfig, 'inner_b': default_qconfig}
+        model = torch.jit.script(Outer())
+        # make sure it runs
+        prepare_jit(model, qconfig_dict)
+
+        model = torch.jit.trace(Outer(), torch.randn(3))
```
Comment (jerryzh168) [inline test/quantization/test_quantize_jit.py:467]: nit: we can put script and trace in a loop and use get_script_module: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L130

you can find example uses of `get_script_module` in this file

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 4]
```diff
@@ -124,11 +124,24 @@ class ModuleCloneHelper {
     size_t N = type->numAttributes();
     for (size_t i = 0; i < N; ++i) {
       IValue s = module._ivalue()->getSlot(i);
-      if (type->getAttribute(i)->is_module()) {
```
Comment (suo) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:127]: Extremely suspicious that we have two highly similar versions of cloning :/, but that is not related to this PR

**Review by suo — APPROVED** [review]
